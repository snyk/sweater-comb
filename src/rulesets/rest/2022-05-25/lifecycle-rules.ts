import { RuleError, Ruleset, SpecificationRule } from "@useoptic/rulesets-base";
import { links } from "../../../docs";
import { stabilityKey } from "./constants";

const stabilityRequirement = new SpecificationRule({
  name: "resource stability",
  docsLink: links.versioning.stabilityLevels,
  matches: (specification, rulesContext) =>
    rulesContext.specification.change !== "removed",
  rule: (specificationAssertions) => {
    specificationAssertions.requirement(
      "be provided for every resource document",
      (specification) => {
        const stability: string | undefined = specification.value[stabilityKey];
        const allowed = ["wip", "experimental", "beta", "ga"];
        if (!stability || !allowed.includes(stability)) {
          throw new RuleError({
            message: `${stability} must be one of allowed values ${allowed.join(
              ", ",
            )}`,
          });
        }
      },
    );
  },
});

const stabilityTransitions = new SpecificationRule({
  name: "resource stability transitions",
  docsLink: links.versioning.promotingStability,
  rule: (specificationAssertions) => {
    specificationAssertions.changed(
      "not change unless it was wip",
      (before, after) => {
        const beforeStability: string | undefined = before.value[stabilityKey];
        const afterStability: string | undefined = after.value[stabilityKey];
        // When there is no `before`, it means it's a new file. Any stability is allowed.
        // When there is no `after`, it means it's been deleted and sunset rules should apply in another rule.
        if (!beforeStability || !afterStability) return;
        // A resource can go from wip to anything, so no need to check.
        if (beforeStability === "wip") return;

        if (beforeStability !== afterStability) {
          throw new RuleError({
            message: `stability transition from '${before}' to '${after}' not allowed`,
          });
        }
      },
    );
  },
});

const followSunsetRules = new SpecificationRule({
  name: "sunset rules",
  matches: (specification, rulesContext) => {
    return (
      !(
        specification.value[stabilityKey] === "wip" ||
        specification.value[stabilityKey] === "experimental"
      ) && rulesContext.specification.change === "removed"
    );
  },
  rule: (specificationAssertions, rulesContext) => {
    specificationAssertions.requirement("follow sunset rules", () => {
      const {
        resourceVersions,
        changeResource: resourceName,
        changeVersion: { date, stability },
        changeDate,
      } = rulesContext.custom;

      const deprecatedBy =
        resourceVersions?.[resourceName]?.[date]?.[stability]?.deprecatedBy;

      if (!deprecatedBy) {
        throw new RuleError({
          message: `expected ${resourceName} to be deprecated before removing`,
        });
      }

      const contextDate = new Date(changeDate);
      const resourceDate = new Date(date);

      const diffDays =
        (contextDate.getTime() - resourceDate.getTime()) /
        (1000 * 60 * 60 * 24);

      // Number of days required before a resource can be removed
      const sunsetSchedule = {
        beta: 90,
        ga: 180,
      };
      const requiredDays = sunsetSchedule[stability];

      if (!requiredDays) {
        throw new RuleError({
          message: `unexpected stability ${stability} in ${resourceName}`,
        });
      }
      if (diffDays < requiredDays) {
        throw new RuleError({
          message: `expected ${stability} resource ${resourceName} to be deprecated ${requiredDays} days`,
        });
      }
    });
  },
});

const noNewExperimental = new SpecificationRule({
  name: "no new experimental versions",
  docsLink: links.versioning.stabilityLevels,
  matches: (specification, rulesContext) =>
    rulesContext.specification.change === "added",
  rule: (specificationAssertions) => {
    specificationAssertions.requirement(
      "be provided for every resource document",
      (specification) => {
        const stability: string | undefined = specification.value[stabilityKey];
        if (!stability) {
          return;
        }
        if (stability === "wip" || stability === "experimental") {
          throw new RuleError({
            message: `new endpoints cannot use ${stability} stability`,
          });
        }
      },
    );
  },
});

export const lifecycleRuleset = new Ruleset({
  name: "api lifecycle ruleset",
  rules: [
    stabilityRequirement,
    stabilityTransitions,
    followSunsetRules,
    noNewExperimental,
  ],
});
