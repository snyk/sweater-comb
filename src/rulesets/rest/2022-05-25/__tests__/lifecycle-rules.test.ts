import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { RuleRunner, TestHelpers } from "@useoptic/rulesets-base";
import { context } from "./fixtures";

import { lifecycleRuleset } from "../lifecycle-rules";
import { stabilityKey } from "../constants";

const baseJson = TestHelpers.createEmptySpec();

describe("lifecycle rules", () => {
  describe("stability", () => {
    test.each(["wip", "experimental", "beta", "ga"])(
      "a valid stability is provided - %s",
      (stability) => {
        const ruleRunner = new RuleRunner([lifecycleRuleset]);
        const ruleInputs = {
          ...TestHelpers.createRuleInputs(baseJson, {
            ...baseJson,
            [stabilityKey]: stability,
          } as OpenAPIV3.Document),
          context,
        };
        const results = ruleRunner.runRulesWithFacts(ruleInputs);

        expect(results.length).toBeGreaterThan(0);
        expect(results.every((result) => result.passed)).toBe(true);
        expect(results).toMatchSnapshot();
      },
    );

    test("an invalid stability is provided", () => {
      const ruleRunner = new RuleRunner([lifecycleRuleset]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, {
          ...baseJson,
          [stabilityKey]: "published",
        } as OpenAPIV3.Document),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);

      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(false);
      expect(results).toMatchSnapshot();
    });

    test("wip can be changed to another maturity", () => {
      const ruleRunner = new RuleRunner([lifecycleRuleset]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(
          {
            ...baseJson,
            [stabilityKey]: "wip",
          } as OpenAPIV3.Document,
          {
            ...baseJson,
            [stabilityKey]: "ga",
          } as OpenAPIV3.Document,
        ),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);

      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(true);
      expect(results).toMatchSnapshot();
    });

    test("can not change from any stability but wip", () => {
      const ruleRunner = new RuleRunner([lifecycleRuleset]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(
          {
            ...baseJson,
            [stabilityKey]: "ga",
          } as OpenAPIV3.Document,
          {
            ...baseJson,
            [stabilityKey]: "beta",
          } as OpenAPIV3.Document,
        ),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);

      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(false);
      expect(results).toMatchSnapshot();
    });
  });

  describe("sunset", () => {
    test("fails when the file was removed and not deprecated", () => {
      const ruleRunner = new RuleRunner([lifecycleRuleset]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, {
          ...baseJson,
          ["x-optic-ci-empty-spec"]: true,
        } as OpenAPIV3.Document),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);
      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(false);
      expect(results).toMatchSnapshot();
    });

    describe("schedule", () => {
      const sunsetContext = {
        changeDate: "2021-10-20",
        changeResource: "Example",
        changeVersion: {
          date: "2021-10-10",
          stability: "beta",
        },
        resourceVersions: {
          Example: {
            "2021-10-10": {
              beta: {
                deprecatedBy: {
                  date: "2021-10-20",
                  stability: "ga",
                },
              },
            },
          },
        },
      };

      test("fails when the schedule isn't met", () => {
        const ruleRunner = new RuleRunner([lifecycleRuleset]);
        const ruleInputs = {
          ...TestHelpers.createRuleInputs(baseJson, {
            ...baseJson,
            ["x-optic-ci-empty-spec"]: true,
          } as OpenAPIV3.Document),
          context: sunsetContext,
        };
        const results = ruleRunner.runRulesWithFacts(ruleInputs);
        expect(results.length).toBeGreaterThan(0);
        expect(results.every((result) => result.passed)).toBe(false);
        expect(results).toMatchSnapshot();
      });

      test("passes when the schedule is met", () => {
        const ruleRunner = new RuleRunner([lifecycleRuleset]);
        const ruleInputs = {
          ...TestHelpers.createRuleInputs(baseJson, {
            ...baseJson,
            ["x-optic-ci-empty-spec"]: true,
          } as OpenAPIV3.Document),
          context: { ...sunsetContext, changeDate: "2022-02-01" },
        };
        const results = ruleRunner.runRulesWithFacts(ruleInputs);
        expect(results.length).toBeGreaterThan(0);
        expect(results.every((result) => result.passed)).toBe(true);
        expect(results).toMatchSnapshot();
      });
    });
  });
});
