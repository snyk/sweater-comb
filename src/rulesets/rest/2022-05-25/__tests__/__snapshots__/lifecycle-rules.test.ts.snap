// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lifecycle rules stability a valid stability is provided - beta 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "beta",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#stability-levels",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "beta",
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules stability a valid stability is provided - experimental 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "experimental",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#stability-levels",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "experimental",
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules stability a valid stability is provided - ga 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "ga",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#stability-levels",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "ga",
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules stability a valid stability is provided - wip 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "wip",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#stability-levels",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "wip",
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules stability an invalid stability is provided 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "published",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#stability-levels",
    "error": "published must be one of allowed values wip, experimental, beta, ga",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "published",
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules stability can not change from any stability but wip 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "beta",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#stability-levels",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "beta",
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "ga",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": "stability transition from '[object Object]' to '[object Object]' not allowed",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": false,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules stability wip can be changed to another maturity 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-snyk-api-stability": "ga",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#stability-levels",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "ga",
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-snyk-api-stability": "wip",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules sunset fails when the file was removed and not deprecated 1`] = `
[
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-optic-ci-empty-spec": true,
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-optic-ci-empty-spec": true,
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "expected Example to be deprecated before removing",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "sunset rules",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules sunset schedule fails when the schedule isn't met 1`] = `
[
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-optic-ci-empty-spec": true,
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-optic-ci-empty-spec": true,
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "expected beta resource Example to be deprecated 90 days",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "sunset rules",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;

exports[`lifecycle rules sunset schedule passes when the schedule is met 1`] = `
[
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
          "x-optic-ci-empty-spec": true,
        },
        "before": {
          "info": {
            "title": "Empty",
            "version": "0.0.0",
          },
          "openapi": "3.1.0",
        },
      },
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
    },
    "condition": undefined,
    "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#promoting-stability-of-a-resource-over-time",
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "resource stability transitions",
    "passed": true,
    "received": undefined,
    "type": "changed",
    "where": "specification",
  },
  {
    "change": {
      "location": {
        "conceptualLocation": {},
        "conceptualPath": [],
        "jsonPath": "",
        "kind": "specification",
      },
      "value": {
        "info": {
          "title": "Empty",
          "version": "0.0.0",
        },
        "openapi": "3.1.0",
        "x-optic-ci-empty-spec": true,
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "sunset rules",
    "passed": true,
    "received": undefined,
    "type": "requirement",
    "where": "specification",
  },
]
`;
